include/ensmallen_bits/gradient_descent/update_policies/delta_bar_delta_update.hpp

```diff
@@ -0,0 +1,166 @@
+ *   title   = {Increased Rates of Convergence Through Learning Rate
+ * Adaptation}, author  = {Jacobs, Robert A.}, journal = {Neural Networks},
+ *   volume  = {1},
+ *   number  = {4},
+ *   pages   = {295--307},
+ *   year    = {1988},
+ *   publisher = {Pergamon}
+ * }
+ */
+class DeltaBarDeltaUpdate
+{
+ public:
+  /**
+   * Construct the DeltaBarDelta update policy with given parameters.
+   *
+   * @param kappa The kappa hyperparameter
+   * @param phi The phi hyperparameter
+   * @param momentum The momentup hyperparameter
+   * @param minGain The minGain hyperparameter
+   */
+  DeltaBarDeltaUpdate(const double kappa = 0.2,
+                      const double phi = 0.8,
+                      const double momentum = 0.5,
+                      const double minGain = 0.01)
+      : kappa(kappa), phi(phi), momentum(momentum), minGain(minGain)
+  {
+    /* Do nothing. */
+  }
+
+  //! Access kappa.
+  double Kappa() const { return kappa; }
+  //! Modify kappa.
+  double& Kappa() { return kappa; }
+
+  //! Access phi.
+  double Phi() const { return phi; }
+  //! Modify phi.
+  double& Phi() { return phi; }
+
+  //! Access the momentum.
+  double Momentum() const { return momentum; }
+  //! Modify the momentum.
+  double& Momentum() { return momentum; }
+
+  //! Access the minGain
+  double MinimumGain() const { return minGain; }
+  //! Modify the minGain.
+  double& MinimumGain() { return minGain; }
+
+  /**
+   * The UpdatePolicyType policy classes must contain an internal 'Policy'
+   * template class with two template arguments: MatType and GradType.  This is
+   * instantiated at the start of the optimization, and holds parameters
+   * specific to an individual optimization.
+   */
+  template <typename MatType, typename GradType>
+  class Policy
+  {
+   public:
+    // typedef typename MatType::elem_type ElemType;
+
+    /**
+     * This is called by the optimizer method before the start of the iteration
+     * update process.
+     *
+     * @param parent Instantiated parent class.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(const DeltaBarDeltaUpdate& parent,
+           const size_t rows,
+           const size_t cols)
+        : parent(parent), velocity(rows, cols)
+    {
+      gains.ones(rows, cols);
+    }
+
+    /**
+     * Update step for Gradient Descent.
+     *
+     * @param iterate Parameters that minimize the function.
+     * @param stepSize Step size to be used for the given iteration.
+     * @param gradient The gradient matrix.
+     */
+    void Update(MatType& iterate,
+                const double stepSize,
+                const GradType& gradient)
+    {
+      const MatType mask = conv_to<MatType>::from((gradient % velocity) < 0.0);
+      
+      gains += mask * parent.Kappa();
+      gains %= mask + (1 - mask) *  parent.Phi();
+      gains.clamp(parent.MinimumGain(), arma::datum::inf);
```

Comment:
```suggestion
      gains.clamp(parent.MinimumGain(), arma::Datum<typename MatType::elem_type>::inf);
```
`MatType` may not always be `arma::mat`, so we need to make sure we use the right type here.